prusti-specs/src/extern_spec_rewriter/common.rs

```diff
@@ -1,27 +1,138 @@
+    let (spec_items, generated_attributes) =
+        generate_spec_and_assertions(prusti_attributes, &stub_method)?;
+
+    // In the generated spec items and the stub method:
+    // - Rewrite associated types
+    // - Rewrite "self" to "_self"
+    let self_type_path = parse_quote_spanned! {self_type.span()=> #self_type };
+
+    let mut stub_method = stub_method.expect_impl_item();
+    stub_method.attrs.extend(generated_attributes);
+    stub_method.rewrite_self_type(&self_type_path, self_type_trait);
+    stub_method.rewrite_receiver(&self_type_path);
+
+    // Set span of generated method to externally specified method for better error reporting
+    syn::visit_mut::visit_impl_item_method_mut(&mut SpanOverrider::new(method_sig_span), &mut stub_method);
+
+    let rewritten_spec_items = spec_items.into_iter().map(|spec_item| {
+        match spec_item {
+            syn::Item::Fn(spec_item_fn) => {
+                let mut spec_item_fn: syn::ImplItemMethod = parse_quote_spanned! {spec_item_fn.span()=>
+                    #spec_item_fn
+                };
+                spec_item_fn.rewrite_self_type(&self_type_path, self_type_trait);
+                spec_item_fn.rewrite_receiver(&self_type_path);
+
+                spec_item_fn
             }
-            TokenTree::Ident(mut ident) => {
-                if ident == "self" {
-                    ident = proc_macro2::Ident::new("_self", ident.span());
+            _ => unreachable!(),
+        }
+    }).collect::<Vec<_>>();
+
+    Ok((stub_method, rewritten_spec_items))
+}
+
+/// Given a method signature with parameters, this function returns all typed parameters
+/// as they were used as arguments for the function call.
+/// # Example
+/// Given some function `fn foo(&self, arg1: i32, arg2: bool)`,
+/// returns `self, arg1, arg2`
+pub trait MethodParamsAsCallArguments {
+    fn params_as_call_args(&self) -> Punctuated<Expr, Token![,]>;
+}
+
+impl<H: HasSignature> MethodParamsAsCallArguments for H {
+    fn params_as_call_args(&self) -> Punctuated<Expr, Token!(,)> {
+        self.sig().inputs.params_as_call_args()
+    }
+}
+
+impl MethodParamsAsCallArguments for Punctuated<FnArg, Token![,]> {
+    fn params_as_call_args(&self) -> Punctuated<Expr, Token!(,)> {
+        let mut args = Punctuated::new();
+        for param in self.iter() {
+            let span = param.span();
+            match param {
+                FnArg::Typed(PatType { pat: box Pat::Ident(ident), .. }) => {
+                    args.push(parse_quote_spanned! {span=>#ident });
+                    args.push_punct(<Token![,]>::default());
+                },
+                FnArg::Receiver(_) => {
+                    args.push(parse_quote_spanned! {span=>self})
                 }
-                ident.into_token_stream()
+                _ => unimplemented!(),
             }
-            _ => token.into_token_stream(),
-        }))
+        }
+        args
```

Comment:
`Punctuated` has [`FromIterator`](https://docs.rs/syn/latest/syn/punctuated/struct.Punctuated.html#impl-FromIterator%3CT%3E), so all this should be expressible with an iterator chain + `map` + `collect`. The `push_punct` in particular seems unneccessary if it is not correctly spanned anyway.